File: src/utils/notebookFormat.ts
export type RegisteredNotebookFormatId = 'blocks' | 'markdown' | 'writer';
export type NotebookFormatId = RegisteredNotebookFormatId | 'unsupported';

export interface NotebookFormatDef {
  id: RegisteredNotebookFormatId;
  /** Canonical compound extension including `.md` for newly created files. */
  extension: string;
  /** Legacy suffixes that still map to this format (e.g. plain `.md` → blocks). */
  aliases?: readonly string[];
  labelKey: string;
}

/**
 * Notebook formats keyed by filename suffix.
 * Longer / more specific suffixes are matched first.
 * Register future formats (treemind / checklist / …) here.
 */
export const NOTEBOOK_FORMATS: readonly NotebookFormatDef[] = [
  { id: 'blocks', extension: '.blk.md', aliases: ['.md'], labelKey: 'directory.formats.blocks' },
  { id: 'markdown', extension: '.mk.md', labelKey: 'directory.formats.markdown' },
  { id: 'writer', extension: '.writer.md', labelKey: 'directory.formats.writer' },
] as const;

const FORMAT_BY_ID = Object.fromEntries(
  NOTEBOOK_FORMATS.map((format) => [format.id, format]),
) as Record<RegisteredNotebookFormatId, NotebookFormatDef>;

/** All known suffixes sorted longest-first for unambiguous matching. */
const SUFFIX_MATCHERS: readonly { format: RegisteredNotebookFormatId; suffix: string }[] = NOTEBOOK_FORMATS
  .flatMap((format) => [
    { format: format.id, suffix: format.extension },
    ...(format.aliases ?? []).map((suffix) => ({ format: format.id, suffix })),
  ])
  .sort((a, b) => b.suffix.length - a.suffix.length);

/** Extra token before `.md`, e.g. `.treemind.md` / `.checklist.md`. */
const COMPOUND_MD_SUFFIX_RE = /(\.[^./\\]+)\.md$/i;

export function getFormatDef(format: NotebookFormatId): NotebookFormatDef {
  if (format === 'unsupported') return FORMAT_BY_ID.blocks;
  return FORMAT_BY_ID[format] ?? FORMAT_BY_ID.blocks;
}

/** Canonical extension used when creating new notebooks of this format. */
export function getFormatExtension(format: NotebookFormatId): string {
  if (format === 'unsupported') return FORMAT_BY_ID.blocks.extension;
  return getFormatDef(format).extension;
}

/**
 * Compound `.token.md` suffix that is not registered in this app version.
 * Plain legacy `.md` and known format markers are not unknown.
 */
export function getUnknownNotebookFormatSuffix(fileName: string): string | null {
  const lower = fileName.toLowerCase();
  if (!lower.endsWith('.md')) return null;

  for (const { suffix } of SUFFIX_MATCHERS) {
    if (suffix === '.md') continue;
    if (lower.endsWith(suffix)) return null;
  }

  const match = COMPOUND_MD_SUFFIX_RE.exec(lower);
  if (!match) return null;
  return fileName.slice(fileName.length - match[0].length);
}

export function getMatchedNotebookSuffix(fileName: string): string | null {
  const unknown = getUnknownNotebookFormatSuffix(fileName);
  if (unknown) return unknown;

  const lower = fileName.toLowerCase();
  for (const { suffix } of SUFFIX_MATCHERS) {
    if (lower.endsWith(suffix)) return fileName.slice(fileName.length - suffix.length);
  }
  return null;
}

export function detectNotebookFormat(fileName: string): NotebookFormatId {
  if (getUnknownNotebookFormatSuffix(fileName)) return 'unsupported';

  const lower = fileName.toLowerCase();
  for (const { format, suffix } of SUFFIX_MATCHERS) {
    if (lower.endsWith(suffix)) return format;
  }
  return 'blocks';
}

export function getNotebookDisplayName(fileName: string): string {
  const suffix = getMatchedNotebookSuffix(fileName);
  if (suffix) {
    return fileName.slice(0, fileName.length - suffix.length);
  }
  return fileName.replace(/\.md$/i, '');
}

export function isDocumentNotebookFormat(format: NotebookFormatId): boolean {
  return format === 'markdown' || format === 'writer';
}

export function isUnsupportedNotebookFormat(format: NotebookFormatId): boolean {
  return format === 'unsupported';
}

/** Markdown editor, including compatibility mode for unknown future suffixes. */
export function shouldUseMarkdownEditor(
  format: NotebookFormatId,
  compatOpenAsMarkdown?: boolean,
): boolean {
  return format === 'markdown' || !!compatOpenAsMarkdown;
}

/** Persist the raw document body instead of serializing note blocks. */
export function shouldSaveAsDocument(
  format: NotebookFormatId,
  compatOpenAsMarkdown?: boolean,
): boolean {
  return isDocumentNotebookFormat(format) || !!compatOpenAsMarkdown;
}

/**
 * Markdown and writer notes share the same document body.
 * Only the filename suffix (and therefore the editor view) differs.
 */
export function getSwappableArticleFormat(
  format: NotebookFormatId,
): Extract<NotebookFormatId, 'markdown' | 'writer'> | null {
  if (format === 'markdown') return 'writer';
  if (format === 'writer') return 'markdown';
  return null;
}

/** Rebuild a notebook filename with a different format suffix, keeping the display name. */
export function replaceNotebookFormatSuffix(
  fileName: string,
  targetFormat: NotebookFormatId,
): string {
  return buildNotebookFileName(getNotebookDisplayName(fileName), targetFormat);
}

export interface ResolvedNotebookFileName {
  fileName: string;
  format: NotebookFormatId;
  displayName: string;
}

export interface ResolveNotebookFileNameOptions {
  /** Keep this exact suffix (e.g. legacy `.md` when renaming an old notebook). */
  preserveExtension?: string | null;
}

/**
 * Resolve a user-entered notebook name into a concrete filename + format.
 * Accepts bare names, compound stems (`notes.blk` / `notes.mk`),
 * or full filenames (`notes.blk.md` / legacy `notes.md`).
 */
export function resolveNotebookFileName(
  inputName: string,
  preferredFormat: NotebookFormatId = 'blocks',
  options?: ResolveNotebookFileNameOptions,
): ResolvedNotebookFileName {
  const trimmed = inputName.trim();
  const preserved = options?.preserveExtension
    ? normalizePreserveExtension(options.preserveExtension, preferredFormat)
    : null;

  if (!trimmed) {
    const extension = preserved ?? getFormatExtension(preferredFormat);
    return {
      fileName: `untitled${extension}`,
      format: preferredFormat,
      displayName: 'untitled',
    };
  }

  if (/\.md$/i.test(trimmed)) {
    const format = detectNotebookFormat(trimmed);
    return {
      fileName: trimmed,
      format,
      displayName: getNotebookDisplayName(trimmed),
    };
  }

  for (const { format, suffix } of SUFFIX_MATCHERS) {
    const stem = suffix.replace(/\.md$/i, '');
    if (!stem || stem === '.') continue;
    if (trimmed.toLowerCase().endsWith(stem)) {
      const displayName = trimmed.slice(0, trimmed.length - stem.length);
      return {
        fileName: `${trimmed}.md`,
        format,
        displayName: displayName || trimmed,
      };
    }
  }

  const extension = preserved ?? getFormatExtension(preferredFormat);
  return {
    fileName: `${trimmed}${extension}`,
    format: preferredFormat,
    displayName: trimmed,
  };
}

function normalizePreserveExtension(
  extension: string,
  preferredFormat: NotebookFormatId,
): string {
  const normalized = extension.startsWith('.') ? extension : `.${extension}`;
  const lower = normalized.toLowerCase();
  if (getUnknownNotebookFormatSuffix(`stem${lower}`)) {
    return normalized;
  }
  for (const { format, suffix } of SUFFIX_MATCHERS) {
    if (suffix === lower && format === preferredFormat) {
      return normalized;
    }
  }
  return getFormatExtension(preferredFormat);
}

export function buildNotebookFileName(displayName: string, format: NotebookFormatId): string {
  return resolveNotebookFileName(displayName, format).fileName;
}

export function isMarkdownNotebookFileName(fileName: string): boolean {
  return /\.md$/i.test(fileName);
}

/**
 * True when the filename already carries a canonical format marker,
 * or an unknown compound `.token.md` suffix that a newer version may own.
 */
export function hasExplicitNotebookFormatMarker(fileName: string): boolean {
  const lower = fileName.toLowerCase();
  if (NOTEBOOK_FORMATS.some((format) => lower.endsWith(format.extension.toLowerCase()))) {
    return true;
  }
  return getUnknownNotebookFormatSuffix(fileName) != null;
}

/**
 * Keep files that already have a built-in format marker.
 * Plain `.md` (no blk / mk / writer marker) is rewritten to `.writer.md`.
 */
export function resolveImportedNotebookFileName(fileName: string): { fileName: string; converted: boolean } {
  if (hasExplicitNotebookFormatMarker(fileName)) {
    return { fileName, converted: false };
  }
  return {
    fileName: buildNotebookFileName(getNotebookDisplayName(fileName), 'writer'),
    converted: true,
  };
}

File: mobile/lib/core/notebook_format.dart
enum NotebookFormat { blocks, markdown, writer, unsupported }

class NotebookFormatDef {
  const NotebookFormatDef({
    required this.id,
    required this.extension,
    this.aliases = const [],
  });

  final NotebookFormat id;
  final String extension;
  final List<String> aliases;
}

/// Notebook formats keyed by filename suffix.
/// Longer / more specific suffixes are matched first.
const notebookFormats = <NotebookFormatDef>[
  NotebookFormatDef(
    id: NotebookFormat.blocks,
    extension: '.blk.md',
    aliases: ['.md'],
  ),
  NotebookFormatDef(id: NotebookFormat.markdown, extension: '.mk.md'),
  NotebookFormatDef(id: NotebookFormat.writer, extension: '.writer.md'),
];

class _SuffixMatcher {
  const _SuffixMatcher(this.format, this.suffix);
  final NotebookFormat format;
  final String suffix;
}

final _formatById = {for (final format in notebookFormats) format.id: format};

/// All known suffixes sorted longest-first for unambiguous matching.
final _suffixMatchers = () {
  final matchers = <_SuffixMatcher>[
    for (final format in notebookFormats) ...[
      _SuffixMatcher(format.id, format.extension),
      ...format.aliases.map((suffix) => _SuffixMatcher(format.id, suffix)),
    ],
  ]..sort((a, b) => b.suffix.length.compareTo(a.suffix.length));
  return List<_SuffixMatcher>.unmodifiable(matchers);
}();

final _compoundMdSuffix = RegExp(r'(\.[^./\\]+)\.md$', caseSensitive: false);

extension NotebookFormatX on NotebookFormat {
  bool get isDocument =>
      this == NotebookFormat.markdown || this == NotebookFormat.writer;

  bool get isUnsupported => this == NotebookFormat.unsupported;

  String get extension =>
      (_formatById[this] ?? _formatById[NotebookFormat.blocks]!).extension;

  NotebookFormat? get swappableArticleFormat {
    if (this == NotebookFormat.markdown) return NotebookFormat.writer;
    if (this == NotebookFormat.writer) return NotebookFormat.markdown;
    return null;
  }
}

bool shouldUseMarkdownEditor(NotebookFormat format, {bool compatOpenAsMarkdown = false}) {
  return format == NotebookFormat.markdown || compatOpenAsMarkdown;
}

bool shouldSaveAsDocument(NotebookFormat format, {bool compatOpenAsMarkdown = false}) {
  return format.isDocument || compatOpenAsMarkdown;
}

NotebookFormatDef getFormatDef(NotebookFormat format) {
  return _formatById[format] ?? _formatById[NotebookFormat.blocks]!;
}

String getFormatExtension(NotebookFormat format) => getFormatDef(format).extension;

/// Compound `.token.md` suffix that is not registered in this app version.
String? unknownNotebookFormatSuffix(String fileName) {
  final lower = fileName.toLowerCase();
  if (!lower.endsWith('.md')) return null;

  for (final matcher in _suffixMatchers) {
    if (matcher.suffix == '.md') continue;
    if (lower.endsWith(matcher.suffix)) return null;
  }

  final match = _compoundMdSuffix.firstMatch(lower);
  if (match == null) return null;
  return fileName.substring(fileName.length - match[0]!.length);
}

String? matchedNotebookSuffix(String fileName) {
  final unknown = unknownNotebookFormatSuffix(fileName);
  if (unknown != null) return unknown;

  final lower = fileName.toLowerCase();
  for (final matcher in _suffixMatchers) {
    if (lower.endsWith(matcher.suffix)) {
      return fileName.substring(fileName.length - matcher.suffix.length);
    }
  }
  return null;
}

NotebookFormat detectNotebookFormat(String fileName) {
  if (unknownNotebookFormatSuffix(fileName) != null) {
    return NotebookFormat.unsupported;
  }
  final lower = fileName.toLowerCase();
  for (final matcher in _suffixMatchers) {
    if (lower.endsWith(matcher.suffix)) return matcher.format;
  }
  return NotebookFormat.blocks;
}

String notebookDisplayName(String fileName) {
  final suffix = matchedNotebookSuffix(fileName);
  if (suffix != null) {
    return fileName.substring(0, fileName.length - suffix.length);
  }
  return fileName.replaceAll(RegExp(r'\.md$', caseSensitive: false), '');
}

String replaceNotebookFormatSuffix(String fileName, NotebookFormat targetFormat) {
  return buildNotebookFileName(notebookDisplayName(fileName), targetFormat);
}

class ResolvedNotebookFileName {
  const ResolvedNotebookFileName({
    required this.fileName,
    required this.format,
    required this.displayName,
  });

  final String fileName;
  final NotebookFormat format;
  final String displayName;
}

/// Resolve a user-entered notebook name into a concrete filename + format.
ResolvedNotebookFileName resolveNotebookFileName(
  String inputName, {
  NotebookFormat preferredFormat = NotebookFormat.blocks,
  String? preserveExtension,
}) {
  final trimmed = inputName.trim();
  final preserved = preserveExtension == null
      ? null
      : _normalizePreserveExtension(preserveExtension, preferredFormat);

  if (trimmed.isEmpty) {
    final extension = preserved ?? getFormatExtension(preferredFormat);
    return ResolvedNotebookFileName(
      fileName: 'untitled$extension',
      format: preferredFormat,
      displayName: 'untitled',
    );
  }

  if (RegExp(r'\.md$', caseSensitive: false).hasMatch(trimmed)) {
    final format = detectNotebookFormat(trimmed);
    return ResolvedNotebookFileName(
      fileName: trimmed,
      format: format,
      displayName: notebookDisplayName(trimmed),
    );
  }

  for (final matcher in _suffixMatchers) {
    final stem = matcher.suffix.replaceAll(RegExp(r'\.md$', caseSensitive: false), '');
    if (stem.isEmpty || stem == '.') continue;
    if (trimmed.toLowerCase().endsWith(stem)) {
      final displayName = trimmed.substring(0, trimmed.length - stem.length);
      return ResolvedNotebookFileName(
        fileName: '$trimmed.md',
        format: matcher.format,
        displayName: displayName.isEmpty ? trimmed : displayName,
      );
    }
  }

  final extension = preserved ?? getFormatExtension(preferredFormat);
  return ResolvedNotebookFileName(
    fileName: '$trimmed$extension',
    format: preferredFormat,
    displayName: trimmed,
  );
}

String _normalizePreserveExtension(String extension, NotebookFormat preferredFormat) {
  final normalized = extension.startsWith('.') ? extension : '.$extension';
  final lower = normalized.toLowerCase();
  if (unknownNotebookFormatSuffix('stem$lower') != null) {
    return normalized;
  }
  for (final matcher in _suffixMatchers) {
    if (matcher.suffix == lower && matcher.format == preferredFormat) {
      return normalized;
    }
  }
  return getFormatExtension(preferredFormat);
}

String buildNotebookFileName(String displayName, NotebookFormat format) {
  return resolveNotebookFileName(displayName, preferredFormat: format).fileName;
}

String initialNotebookContent(NotebookFormat format, String displayName) {
  if (format.isDocument) {
    return '# $displayName\n\n';
  }
  final now = DateTime.now().toUtc().toIso8601String();
  return '---\n'
      'title: $displayName\n'
      'tags: []\n'
      'createdAt: $now\n'
      'updatedAt: $now\n'
      '---\n\n';
}

bool isMarkdownNotebookFileName(String fileName) {
  return RegExp(r'\.md$', caseSensitive: false).hasMatch(fileName);
}

bool hasExplicitNotebookFormatMarker(String fileName) {
  final lower = fileName.toLowerCase();
  if (notebookFormats.any(
    (format) => lower.endsWith(format.extension.toLowerCase()),
  )) {
    return true;
  }
  return unknownNotebookFormatSuffix(fileName) != null;
}

/// Keep files that already have a built-in format marker.
/// Plain `.md` (no blk / mk / writer marker) is rewritten to `.writer.md`.
({String fileName, bool converted}) resolveImportedNotebookFileName(String fileName) {
  if (hasExplicitNotebookFormatMarker(fileName)) {
    return (fileName: fileName, converted: false);
  }
  return (
    fileName: buildNotebookFileName(
      notebookDisplayName(fileName),
      NotebookFormat.writer,
    ),
    converted: true,
  );
}

class NotebookExistsException implements Exception {
  const NotebookExistsException(this.name);
  final String name;
}

class NotebookFormatNotSwappableException implements Exception {
  const NotebookFormatNotSwappableException();
}
